pkg/apis/workflow/v1alpha1/workflow_types.go

```diff
@@ -1309,26 +1401,29 @@ func (w *Workflow) SearchArtifacts(q *ArtifactSearchQuery) ArtifactSearchResults
 	var results ArtifactSearchResults
 
 	for _, n := range w.Status.Nodes {
+		if q.TemplateName != "" && n.TemplateName != q.TemplateName {
+			continue
+		}
+		if q.NodeId != "" && n.ID != q.NodeId {
+			continue
+		}
+		if q.NodeTypes != nil && !q.NodeTypes[n.Type] {
+			continue
+		}
```

Comment:
moved the TemplateName and NodeId tests up here since they're node level and don't need to be performed for every artifact on the node